### ISSUE TITLE
Swagger supports normal minimal asp.net core apis but not work for Giraffe endpoints

### ISSUE THREAD

**Issue by albertwoo** [giraffe-fsharp/Giraffe#488]
I am playing around with asp.net core 6.0.100-preview.6.21355.2
I found the swagger works for default endpoint routes like in below code the /hello2 will be picked by swagger, but not work with giraffe endpoints.

```fsharp
open System
open Microsoft.AspNetCore.Http
open Microsoft.AspNetCore.Builder
open Microsoft.Extensions.Hosting
open Microsoft.Extensions.DependencyInjection
open Giraffe
open Giraffe.EndpointRouting

let builder = WebApplication.CreateBuilder()

builder.Services.AddEndpointsApiExplorer()
builder.Services.AddHttpContextAccessor() |> ignore
builder.Services.AddSwaggerGen() |> ignore

let application = builder.Build()

application.UseSwagger() |> ignore
application.MapGiraffeEndpoints
    [
        GET [
            route "/hello1" (text "Hello")
        ]
    ]
application.MapGet("/hello2", Func<_, _>(fun (ctx: HttpContext) -> $"Hi {ctx.Request.Path}")) |> ignore
application.UseSwaggerUI() |> ignore

application.Run()
```

**Comment by albertwoo** [comment]
This root reason should be similar with falco:
https://github.com/pimbrouwers/Falco/issues/49

**Comment by Numpsy** [comment]
This was mentioned in the talk at https://www.twitch.tv/videos/1092464368 yesterday (right at the end, and in relation to https://github.com/baronfel/Giraffe.EndpointRouting.OpenAPI I think).
It sounds from that like the requirements for the new MS stuff to work on its own are quite specific (I haven't tried using the new v6 stuff myself yet though)

**Comment by baronfel** [comment]
Yep I brought it up on purpose because I'd seen this discussion pop up again (and also I'm the Falco repo).

The relevant code in aspnetcore is [here](https://github.com/dotnet/aspnetcore/blob/3accbd6abe9701c431855a5c4ab5b6098841cf5d/src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs), the gist of which is to translate specifically minimal endpoints that are described by lambdas (or method groups) by their endpoint Metadata (which includes the reflection Method info) into ApiDescriptors. 

This doesn't work for giraffe at all because composed endpoints don't match this pattern, so I'd we want to interop at the level that Swashbuckler does, we'd need to implement our own Apidescriptor provider that could create these same structures from giraffe apis.

That's where the fiddly work comes in, as far as I can see it.

**Comment by jkone27** [comment]
this project was archived but could be helpful, at least for the openapi CE writer part?
https://github.com/akhansari/FSharp.OpenApi

**Comment by jkone27** [comment]
https://github.com/dotnet/aspnetcore/issues/37098

**Comment by jkone27** [comment]
maybe something new can be tried with these updates on openapi and metadata?

https://devblogs.microsoft.com/dotnet/asp-net-core-updates-in-dotnet-7-preview-4/

**Comment by voronoipotato** [comment]
This would be really helpful for companies that are reliant on openapi and swagger. @baronfel  thanks for the work to describe this.

**Comment by captainsafia** [comment]
Coming across this issue from https://github.com/domaindrivendev/Swashbuckle.AspNetCore/issues/2300.

> This doesn't work for giraffe at all because composed endpoints don't match this pattern, so if we want to interop at the level that Swashbuckle does, we'd need to implement our own Apidescriptor provider that could create these same structures from giraffe apis. Swashbuckle would then be able to take our ApiDescription and generate the swagger doc from that.

@baronfel's recommendation here is exactly right. Providing a custom `IApiDescriptionProvider` interface is the pathway that frameworks should take in order to light up support for OpenAPI in ASP.NET Core. The `IApiDescriptionProvider` is essentially the middleman introspection layer that allows other tooling (like Swashbuckle, NSwag, or the new built-in OpenAPI support coming in .NET 9) to determine what metadata it should generate.

The link above points to minimal API's implementation (see [here](https://github.com/dotnet/aspnetcore/blob/ad4c5fe13abd1a8fb755e3cdc56001af4e49bfdb/src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs)). There's also implementations for MVC (see [here](https://github.com/dotnet/aspnetcore/blob/ad4c5fe13abd1a8fb755e3cdc56001af4e49bfdb/src/Mvc/Mvc.ApiExplorer/src/DefaultApiDescriptionProvider.cs)) and gRPC (see [here](https://github.com/dotnet/aspnetcore/blob/ad4c5fe13abd1a8fb755e3cdc56001af4e49bfdb/src/Grpc/JsonTranscoding/src/Microsoft.AspNetCore.Grpc.Swagger/Internal/GrpcJsonTranscodingDescriptionProvider.cs)) that can be used as examples.

**Comment by voronoipotato** [comment]
@captainsafia  thank you so much for helping provide tangible examples. maybe one day if I can find the time I'll try and mess around with an implementation attempt. Having swagger for Giraffe would really help unify our C# and F# api descriptions.

**Comment by 64J0** [comment]
Apparently it was fixed by PR https://github.com/giraffe-fsharp/Giraffe/pull/599. What you think?

**Comment by captainsafia** [comment]
> Apparently it was fixed by PR #599. What you think?

One thing about this approach I spotted is that it makes use of the `WithOpenApi` extension method to opt-in to the OpenAPI generation done [by this bit of code](https://github.com/dotnet/aspnetcore/blob/662d200bc4feb11a01895e675fbfee8517c6fe2a/src/OpenApi/src/Services/OpenApiGenerator.cs). One thing to note about this pattern is that it only works with Swashbuckle and there's no plans to support it in the new Microsoft.AspNetCore.OpenApi package. See https://github.com/domaindrivendev/Swashbuckle.AspNetCore/issues/2849.

One thing I'm observing about your implementation though: do the endpoints generated by Giraffe eventually get populated into the `EndpointDataSource`? If so, then the built-in `IApiDescriptionProvider` implementation for endpoints (see [here](https://github.com/dotnet/aspnetcore/blob/662d200bc4feb11a01895e675fbfee8517c6fe2a/src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs)) should be able to pick them up and you'll be able to light-up OpenAPI support without taking a dependency on a singular package (since they all use ApiDescription's as the source of truth).

**Comment by NinoFloris** [comment]
It's likely that they're added as RequestDelegates instead of taking the Delegate overload. Only the latter will populate the MethodInfo required to make them flow through automatically through the built-in `IApiDescriptionProvider`. Ran into exactly this recently while fixing a closed source 'giraffe light' implementation. F# doesn't have the automatic coercion from lambdas (FSharpFunc) to Delegate like C# introduced in 10.0 which meant explicit RequestDelegate construction spread all over the F# ecosystem, even when it was the only endpoint type.

**Comment by jkone27** [comment]
Maybe related for reference 

* https://github.com/Lanayx/Oxpecker/issues/25

* https://github.com/pimbrouwers/Falco/tree/develop/src/Falco.OpenApi

**Comment by mrtz-j** [comment]
> Maybe related for reference
> 
>     * [openapi not working (tried both builder and slimbuilder) Lanayx/Oxpecker#25](https://github.com/Lanayx/Oxpecker/issues/25)
> 
>     * https://github.com/pimbrouwers/Falco/tree/develop/src/Falco.OpenApi

I've a working sample app [here](https://github.com/giraffe-fsharp/Giraffe.OpenApi). It uses the `Giraffe.OpenApi` package.

**Comment by captainsafia** [comment]
> Does anything need to get merged or is this currently working?

I think this currently works with the caveat that it uses the soon-to-be-deprecated-ish `WithOpenApi` extension method in ASP.NET Core which is currently only designed to work with Swashbuckle (see https://github.com/giraffe-fsharp/Giraffe/issues/488#issuecomment-2237512214). If you're OK with that constraint, then you can probably consider this resolved with the implementation in the package referenced above.
